### ISSUE TITLE
Letsencrypt commands are pointed to ZeroSSL by default

### ISSUE THREAD

**Issue by nova0052** [WordOps/WordOps#387]
commands referencing lets encrypt, e.g. ```wo site update example.com -le```, are sending requests to ZeroSSL by default instead of LetsEncrypt.

The problem is caused by line 32 in /etc/letsencrypt/acme.sh, where the default CA is set to ZeroSSL instead of LetsEncrypt.

**Comment by tersor** [comment]
Acme.sh v3.0.0 is using ZeroSSL as default CA. If you run into any issues in WordOps creating new certs, you have two options:

1. Roll with Acme and register a ZeroSSL account:

```
acme.sh --register-account -m my@example.com
```

2. Tell Acme to use Letsencrypt as default CA:

```
acme.sh --set-default-ca --server letsencrypt
```

**Comment by juanpvh** [comment]
@tersor what would be the best option? keep letsencrypt or use ZeroSSL?

**Comment by vikas5914** [comment]
@juanpvh since ZeroSSL only allow 3 certificates for a free account, I think Letsencrypt is better options. 

Personally, I just use the Cloudflare SSL for all my projects.

**Comment by nova0052** [comment]
> Acme.sh v3.0.0 is using ZeroSSL as default CA. If you run into any issues in WordOps creating new certs, you have two options:
> 
> 1. Roll with Acme and register a ZeroSSL account:
> 
> ```
> acme.sh --register-account -m my@example.com
> ```
> 
> 1. Tell Acme to use Letsencrypt as default CA:
> 
> ```
> acme.sh --set-default-ca --server letsencrypt
> ```

@tersor I ended up doing both of these—registration as in #1 above seemed to work fine, but the script still fails to get a cert from ZeroSSL after registration. Log shows the same error after registration as before. I can open a separate issue for that if needed.

Still, it's very counterintuitive that I should have to do this at all; when I enter a command that references LetsEncrypt by name, there's no reason for it to contact ZeroSSL at all.

Is acme.sh part of the WordOps codebase? or is it coming from somewhere else?

**Comment by tersor** [comment]
> since ZeroSSL only allow 3 certificates for a free account, I think Letsencrypt is better options.

@vikas5914 Please correct me if I am wrong, but ZeroSSL offers unlimited "90-Day ACME Certs" [for free](https://zerossl.com/pricing/).

> Is acme.sh part of the WordOps codebase? or is it coming from somewhere else?

@nova0052 Yes, but WordOps only depends on [acme.sh](https://github.com/acmesh-official/acme.sh) if you [use the `--letsencrypt` param](https://docs.wordops.net/commands/site/#lets-encrypt).

Personally, I think acme.sh is great and use it on other non-WordOps servers.

**Comment by vikas5914** [comment]
@tersor Oh yes. the 3 90 days certificates are different.

**Comment by Sl1nkman** [comment]
I have found the same


> Acme.sh v3.0.0 is using ZeroSSL as default CA. If you run into any issues in WordOps creating new certs, you have two options:
> 
> 1. Roll with Acme and register a ZeroSSL account:
> 
> ```
> acme.sh --register-account -m my@example.com
> ```
> 
> 1. Tell Acme to use Letsencrypt as default CA:
> 
> ```
> acme.sh --set-default-ca --server letsencrypt
> ```

This makes no difference, WO CLI is still not issuing certs

Same error no matter what: " 
[Tue Jun 15 14:21:29 UTC 2021] No EAB credentials found for ZeroSSL, let's get one
[Tue Jun 15 14:21:29 UTC 2021] acme.sh is using ZeroSSL as default CA now.
[Tue Jun 15 14:21:29 UTC 2021] Please update your account with an email address first.
[Tue Jun 15 14:21:29 UTC 2021] acme.sh --register-account -m my@example.com
[Tue Jun 15 14:21:29 UTC 2021] See: https://github.com/acmesh-official/acme.sh/wiki/ZeroSSL.com-CA
,
" 
Is there a way to roll back acme.sh in the meantime?

**Comment by vikas5914** [comment]
@Sl1nkman  Can you do the following? 

go to `/opt/acme.sh`  folder and run `git reset --hard a438c84` 

This should change the Letsencrypt as default.

**Comment by nova0052** [comment]
> > Is acme.sh part of the WordOps codebase? or is it coming from somewhere else?
> 
> @nova0052 Yes, but WordOps only depends on [acme.sh](https://github.com/acmesh-official/acme.sh) if you [use the `--letsencrypt` param](https://docs.wordops.net/commands/site/#lets-encrypt).
> 
> Personally, I think acme.sh is great and use it on other non-WordOps servers.

I agree, acme.sh is great and I'm glad it's part of WordOps. The problem is that when I explicitly type ```--letsencrypt```, I expect the script to use LetsEncrypt, not ZeroSSL. 

Using a different CA as default is unintuitive and confusing, especially in this case where ZeroSSL appears not to be working and the script is spitting out errors— _"I need to register my email again? Did LetsEncrypt lose it? Oh, it's using ZeroSSL; that's weird, I thought I told it to use LetsEncrypt, let's try again with LE. Oh no, it's doing ZeroSSL again...WHY? It's using ZeroSSL as default? How do I set it back to LetsEncrypt????"_

It would make sense for ZeroSSL to be the default if the user were entering ```--zerossl``` as the parameter, but that is not the case. We are using the parameter that explicitly says to use Letsencrypt, so the script should be doing so.

**Comment by vikas5914** [comment]
@nova0052 You may need to `--letsencrypt` inside the WordOps codebase where its calling acme.sh

**Comment by nova0052** [comment]
> @nova0052 You may need to `--letsencrypt` inside the WordOps codebase where its calling acme.sh

Sorry, I don't understand your meaning. 

I'm already using the ```--letsencrypt``` parameter, the problem is that the parameter is ignored and acme.sh is using ZeroSSL instead of LetsEncrypt.

**Comment by JPaulMora** [comment]
Having this problem too, fixed temporarily by changing this line:
https://github.com/WordOps/WordOps/blob/68cc21aa84b1eb02216d0c9e2e5dae06f3fe2d63/wo/core/acme.py#L94
to this:
```bash
"--issue -d '{0}' {1} -k {2} -f --server letsencrypt"
```

This will fix certificate creation and as far as I checked renewals too. Hope this helps

**Comment by renatofrota** [comment]
> > @nova0052 You may need to `--letsencrypt` inside the WordOps codebase where its calling acme.sh
> 
> Sorry, I don't understand your meaning.
> 
> I'm already using the `--letsencrypt` parameter, the problem is that the parameter is ignored and acme.sh is using ZeroSSL instead of LetsEncrypt.

WordOps developers could not predict acme.sh would someday change it's default CA from Let's Encrypt to ZeroSSL. Money talks, apilayer (Idera, Inc) acquired ZeroSSL and also acquired (or just partnered?) acme.sh developers and ZeroSSL is now the default CA of acme.sh script.

The WordOps parameter "--letsencrypt" should have been named "--acmesh" or something like that. Maybe it's too late to change. Probably WordOps will set acme.sh default CA to letsencrypt, what is very easy (`acme.sh --set-default-ca --server letsencrypt`) and works flawlessly. Those who have problems with LE certificate issuance rate limit (which is not a problem for me) may change the default to ZeroSSL manually.

**Comment by vilm725** [comment]
Hi every one 😊👋🏼

I've a solution, but not the best. Parameter `acme.sh --set-default-ca --server letsencrypt` or `--letsencrypt`

You can therefore modify the file
`sudo nano /etc/letsencrypt/acme.sh`

At the ligne 35, change :
`DEFAULT_CA=$CA_ZEROSSL`
With 
`DEFAULT_CA=$CA_LETSENCRYPT_V2`

The defaut ssl certifcat with generat by Letsencrypt ✅

It's not the best method : an updat can break this file, but it's work ... If you have a best solution to use Letsencrypt by defaut, do not hesitate to share it whith me 👋🏼

@renatofrota 

> WordOps developers could not predict acme.sh would someday change it's default CA from Let's Encrypt to ZeroSSL. Money talks, apilayer (Idera, Inc) acquired ZeroSSL and also acquired (or just partnered?) acme.sh developers and ZeroSSL is now the default CA of acme.sh script.

Business ...

**Comment by JPaulMora** [comment]
@vilm725 that seems to me like a cleaner solution compared to mine. But that's gonna have to be relayed to the acme.sh team whom I'd think they already chose what's gonna be the default server

**Comment by renatofrota** [comment]
> It's not the best method : an updat can break this file, but it's work ... If you have a best solution to use Letsencrypt by defaut, do not hesitate to share it whith me 👋🏼

Running `acme.sh --set-default-ca --server letsencrypt` **only once** is enough. This is a permanent setting (unless you change it to other CA **manually** afterwards). I have no idea why you (and a lot of people here) are suggesting editing files manually.

**Comment by nova0052** [comment]
> > It's not the best method : an updat can break this file, but it's work ... If you have a best solution to use Letsencrypt by defaut, do not hesitate to share it whith me 👋🏼
> 
> Running `acme.sh --set-default-ca --server letsencrypt` **only once** is enough. This is a permanent setting (unless you change it to other CA **manually** afterwards). I have no idea why you (and a lot of people here) are suggesting editing files manually.

We are editing files manually because we don't know about this command. From the user's perspective, we told WordOps to use LetsEncrypt, and it went off and used ZeroSSL for some unknown reason. 

I suppose a message could be added—"Warning! When you use the --letsencrypt parameter, WordOps will use ZeroSSL instead. To change this behavior, run `acme.sh --set-default-ca --server letsencrypt`"

A better solution is for WordOps to run this command during initial setup+installation. That way, when the end user eventually enters a command with the ```--letsencrypt``` parameter, acme.sh _actually uses letsencrypt_  instead of the current confusing behavior.

**Comment by renatofrota** [comment]
> We are editing files manually because we don't know about this command.

This command was provided to you by @tersor in the first reply of this issue. You just overlooked it.

> From the user's perspective, we told WordOps to use LetsEncrypt, and it went off and used ZeroSSL for some unknown reason.

As I have said, maybe it was not a good choice to use `--letsencrypt`. Using `--acmesh` would be better and less confusing in face of this situation. However, no one could predict the movement of acme.sh changing it's default from Let's Encrypt to ZeroSSL.

> I suppose a message could be added—"Warning! When you use the --letsencrypt parameter, WordOps will use ZeroSSL instead. To change this behavior, run `acme.sh --set-default-ca --server letsencrypt`"
> 
> A better solution is for WordOps to run this command during initial setup+installation. That way, when the end user eventually enters a command with the `--letsencrypt` parameter, acme.sh _actually uses letsencrypt_ instead of the current confusing behavior.

Well said. The parameter could also be changed to --acmesh. This may also be confusing for some time due to the existing documentation, github issues, forum posts, etc mentioning --letsencrypt but is more correct, considering acme.sh is the script called to issue the certificates, no matter which CA you choose.

Yet, the parameter could be just --ssl (acme.sh script itself could have a substitute in the future).

**Comment by nova0052** [comment]
#

> > We are editing files manually because we don't know about this command.
> 
> This command was provided to you by @tersor in the first reply of this issue. You just overlooked it.
Yes, I know about this command thanks to the helpful users here. However, I had already edited config files manually before opening this issue, _because I didn't know about this command._ Other users may experience similar confusion, and it is unreasonable to expect them to know that they must run a special acme.sh command in order for `--letsencrypt` to actually use LetsEncrypt.

> 
> > From the user's perspective, we told WordOps to use LetsEncrypt, and it went off and used ZeroSSL for some unknown reason.
> 
> As I have said, maybe it was not a good choice to use `--letsencrypt`. Using `--acmesh` would be better and less confusing in face of this situation. However, no one could predict the movement of acme.sh changing it's default from Let's Encrypt to ZeroSSL.
> 
Agree, and that is why I opened this issue in the first place, so the confusing behavior could be fixed.

> > I suppose a message could be added—"Warning! When you use the --letsencrypt parameter, WordOps will use ZeroSSL instead. To change this behavior, run `acme.sh --set-default-ca --server letsencrypt`"
> > A better solution is for WordOps to run this command during initial setup+installation. That way, when the end user eventually enters a command with the `--letsencrypt` parameter, acme.sh _actually uses letsencrypt_ instead of the current confusing behavior.
> 
> Well said. The parameter could also be changed to --acmesh. This may also be confusing for some time due to the existing documentation, github issues, forum posts, etc mentioning --letsencrypt but is more correct, considering acme.sh is the script called to issue the certificates, no matter which CA you choose.
> 
> Yet, the parameter could be just --ssl (acme.sh script itself could have a substitute in the future).

I've opened a pull request that will run `acme.sh --set-default-ca --server letsencrypt` during installation, so that moving forward, when a new user types `--letsencrypt`, acme.sh will actually use LetsEncrypt.

**Comment by JPaulMora** [comment]
@renatofrota I forgot to specify this on my first comment, but I went the ugly way because `--set-default-ca` didn't work for me.

**Comment by vikas5914** [comment]
> @renatofrota I forgot to specify this on my first comment, but I went the ugly way because `--set-default-ca` didn't work for me.

what's the value of `DEFAULT_ACME_SERVER` in `etc/letsencrypt/config/account.conf`

**Comment by phlbnks** [comment]
Wow this is fun... Some more evidence / reports for you:

Creating a new site / cert for the first time in ages and cert issuing is broken. Issuing via ZeroSSL, found the log to tell me to register `/etc/letsencrypt/acme.sh --register-account -m xxx@example.com`, however even though that command succeeds, the cert issue still fails with the same error.

I just ran `/etc/letsencrypt/acme.sh --set-default-ca --server letsencrypt` which succeeded, but cert issuing still fails and log shows it's still trying ZeroSSL. 

`etc/letsencrypt/config/account.conf` has no `DEFAULT_ACME_SERVER` value. @vikas5914 can you tell me how this should be defined so I can try setting it manually please?

**Comment by vikas5914** [comment]
@phlbnks are you running the command as `root` or with `sudo`? if not, please try with Sudo and then check.

**Comment by phlbnks** [comment]
Thanks for your reply @vikas5914 - I've tried both:
- `/etc/letsencrypt/acme.sh --register-account -m xxx@example.com` & `/etc/letsencrypt/acme.sh --set-default-ca --server letsencrypt`
- and `sudo /etc/letsencrypt/acme.sh --register-account -m xxx@example.com --force` & `/etc/letsencrypt/acme.sh --set-default-ca --server letsencrypt --force`

The commands all succeed, but `/etc/letsencrypt/config/account.conf` remains free of `DEFAULT_ACME_SERVER` and a script issue still tries to use ZeroSSL and fails with missing registration

**Comment by vikas5914** [comment]
@phlbnks add the following line to `account.conf` and try generate certificate

```
DEFAULT_ACME_SERVER='https://acme-v02.api.letsencrypt.org/directory'
```

**Comment by phlbnks** [comment]
Thanks @vikas5914! That's succeeded in issuing a cert now. Maybe it was just some weirdness with my system, but if I can help test anything or extract any logs to help just let me know.
Thanks again ✨

**Comment by renatofrota** [comment]
> Thanks for your reply @vikas5914 - I've tried both:
> 
>     * `/etc/letsencrypt/acme.sh --register-account -m xxx@example.com` & `/etc/letsencrypt/acme.sh --set-default-ca --server letsencrypt`
> 
>     * and `sudo /etc/letsencrypt/acme.sh --register-account -m xxx@example.com --force` & `/etc/letsencrypt/acme.sh --set-default-ca --server letsencrypt --force`
> 
> 
> The commands all succeed, but `/etc/letsencrypt/config/account.conf` remains free of `DEFAULT_ACME_SERVER` and a script issue still tries to use ZeroSSL and fails with missing registration

You have not prepended command `/etc/letsencrypt/acme.sh --set-default-ca --server letsencrypt` with `sudo`, that was the issue. You have changed the default CA for your own user only. The command has no practical effect since sites are created/updated by WordOps as root - or using `sudo wo`.

**Whenever** you are running a command related to **global** changes on a server it is expected you are doing it as root or using sudo.

For the last time on this thread (I hope) and to help other people coming here due to the same issue:

# How to set acme.sh to use Let's Encrypt as default CA (back from ZeroSSL)

1. Run `acme.sh --set-default-ca --server letsencrypt` -- prepend with `sudo` if you are not running it as root;
2. Done! It is a "set and forget" setting and nothing else needs to be changed.

# How to register at ZeroSSL and use it as default CA

1. (only if you have changed to LE manually) run `acme.sh --set-default-ca --server zerossl` -- prepend with `sudo` if you are not running it as root;
2. Run `acme.sh --register-account -m my@example.com` -- prepend with `sudo` if you are not running it as root;
3. Done! You should now be able to issue certs using ZeroSSL (the WordOps parameter remains `--letsencrypt` or `-le`).

**Comment by vikas5914** [comment]
@renatofrota running with error will give you an error like this 

```
ubuntu@ip-172-26-6-141:~$ sudo /etc/letsencrypt/acme.sh --set-default-ca --server letsencrypt
It seems that you are using sudo, please read this link first:
https://github.com/acmesh-official/acme.sh/wiki/sudo
```

So I think the command should be like this 

```
sudo /etc/letsencrypt/acme.sh --config-home '/etc/letsencrypt/config'  --set-default-ca --server letsencrypt --force

**Comment by phlbnks** [comment]
Thanks for your input @renatofrota and @vikas5914 

I want to confirm that the correct command, tested and working switching between both providers is:

`sudo /etc/letsencrypt/acme.sh --config-home '/etc/letsencrypt/config'  --set-default-ca --server letsencrypt --force`

I'm not just repeating @vikas5914 - I want to highlight that not only is `--force` needed, but also `--config-home '/etc/letsencrypt/config'`. The latter is why the command was failing for me and others earlier - without it the /etc/letsencrypt/config/account.conf file is not modified (the commend exists successfully but has no effect).

There is still the question of why `sudo /etc/letsencrypt/acme.sh --register-account -m xxx@example.com` was (successful exit code but when trying to issue) failing - I wonder if adding `--config-home` will fix it but haven't tested yet.

**Comment by renatofrota** [comment]
Thanks @vikas5914 , updated above.

**Comment by vikas5914** [comment]
> Thanks @vikas5914, updated above.

I think you need to provide --config home in every command since it was changed by WordOps.
